lib/grandpa/grandpa.go

```diff
@@ -547,62 +545,134 @@ func (s *Service) playGrandpaRound() error {
 		})
 	}()
 
+	time.Sleep(interval)
+
 	err = s.attemptToFinalize()
 	if err != nil {
-		log.Error("failed to finalise", "error", err)
+		logger.Error("failed to finalise", "error", err)
 		return err
 	}
 
-	finalised = true
+	close(roundComplete[0])
+	close(roundComplete[1])
```

Comment:
as the two channels are updated at the same time we could use just one channel and share it with the `sendVoteMessage` function

Reply:
`sendVoteMessage` is called twice, once for prevotes and precommits, so there needs to be 2 channels

Reply:
Since the condition to stop the `sendVoteMessage` is the `close(channel)` and both channels are being closed at the same time I think it's possible to create just one channel and
```go
ch := make(chan struct{})
```
start the first and second `sendVoteMessage`goroutines with the same channel 
```go
go s.sendVoteMessage(prevote, vm, ch)
...
go s.sendVoteMessage(precommit, pcm, ch)
```
and then close the channel once
```go
close(ch)
```
This way the 2 goroutines will be stoped as before but sharing just one channel

Reply:
Since the condition to stop the sendVoteMessage is the close(channel) and both channels are being closed at the same time I think it's possible to create just one channel and

```go 
ch := make(chan struct{})
```
start the first and second sendVoteMessagegoroutines with the same channel

```go
go s.sendVoteMessage(prevote, vm, ch)
...
go s.sendVoteMessage(precommit, pcm, ch)
```
and then close the channel once
```go
close(ch)
```
This way the 2 goroutines will be stoped as before but sharing just one channel

Reply:
I've updated this, I was under the impression that closing a channel would only send 1 instance of `struct{}{}` (or whatever the empty type is for the channel) through the channel, but it seems like reading from the closed channel always receives `struct{}{}`